[Merged by Bors] - feat(linear_algebra/general_linear_group): Add some lemmas about SL to GL_pos coercions.

CBirkbeck opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

CBirkbeck committed:
add some lemmas about map from SL to GL_pos

CBirkbeck committed:
commit correct thing now

CBirkbeck committed:
some lint

eric-wieser committed:
style cleanups

eric-wieser reviewed (commented):
I pushed some style cleanups to make this easier to read.

I feel like there's an awful mess of coercions being made here, and the thing to do is to write simp lemmas to standardize the coercions, rather than lemmas about every possible coercion.

I think the lemma you need most is:
```
@[simp] lemma coe_fn_eq_coe (A : GL_pos n R) : ⇑A = (↑(↑A : GL n R) : matrix n n R) := rfl
```
which gets rid of the "dangerous" `has_coe_to_fun` instances that you don't have any lemmas about.

eric-wieser committed:
feat(linear_algebra/matrix.determinant): add `matrix.det_neg`

CBirkbeck committed:
add suggested lemma

eric-wieser committed:
chore(linear_algebra/general_linear_group): clean up accident coe_fn lemmas

This way, all the lemmas are expressed in terms of `↑` and not `⇑`.
This matches the approach used in `special_linear_group`.

eric-wieser commented on src/linear_algebra/general_linear_group.lean:
```diff
@@ -191,6 +193,22 @@ lemma to_GL_pos_injective :
+
+@[simp]
+lemma coe_coe_matrix (g : special_linear_group n ℤ) (i j : n) :
+  ((g : special_linear_group n R) : GL_pos n R) i j = (g : matrix n n ℤ) i j := rfl
```
I think what you actually want here is:
```lean
/-- Coercing a `special_linear_group` to a matrix via `GL_pos` and `GL` is the same as coercing directly. -/
@[simp]
lemma coe_GL_pos_coe_GL_coe_matrix (g : special_linear_group n R) :
  (↑(↑(↑(g : special_linear_group n R) : GL_pos n R) : GL n R) : matrix n n R) = ↑g := rfl
```
This lemma makes your lemma follow trivially

CBirkbeck replied:
Aha yes! I see what you mean now. Ok I've added this and removed the other lemma. Should I merge #12397 into this PR and then wait until that is added?

CBirkbeck replied:
So I merged #12397 into this but it not happy with `coe_to_GL_pos_ext`  and `coe_to_GL_pos_det`.  it keeps saying "function expected at
  g
term has type
  {A // A.det = 1}"

Any idea how to fix this?

CBirkbeck replied:
ok I made an ugly fix to this, but there is probably not the right thing to do.

CBirkbeck committed:
use better lemma

CBirkbeck committed:
lint fix

CBirkbeck committed:
typo in example description

CBirkbeck committed:
fix modular file

CBirkbeck committed:
lint fix

CBirkbeck committed:
re-add lemma

CBirkbeck committed:
Merge remote-tracking branch 'origin/eric-wieser/remove-GL-coe-to-fun' into SL_to_GL_pos_lemmas

CBirkbeck committed:
Merge remote-tracking branch 'origin/master' into SL_to_GL_pos_lemmas

CBirkbeck committed:
merge broke things, need help

CBirkbeck committed:
ugly fix

eric-wieser committed:
Merge branch 'master' into SL_to_GL_pos_lemmas

eric-wieser commented on src/linear_algebra/general_linear_group.lean:
```diff
@@ -186,11 +186,28 @@ lemma to_GL_pos_injective :
  from subtype.coe_injective).of_comp
 
+lemma coe_to_GL_pos_ext (g : special_linear_group n R) (i j : n) :
+  g.1 i j = (g : (GL_pos n R)).1.1 i j := rfl
```
This looks like it's just a backwards version of `coe_GL_pos_coe_GL_coe_matrix` below (after letting `dsimp` replace all the `.1`s with `coe`), so I'd be inclined to remove it.

CBirkbeck replied:
Yes the .1's was just because after I merged with #12397 things had broken.

eric-wieser commented on src/linear_algebra/general_linear_group.lean:
```diff
@@ -186,11 +186,28 @@ lemma to_GL_pos_injective :
+  g.1 i j = (g : (GL_pos n R)).1.1 i j := rfl
+
+lemma coe_to_GL_pos_det (g : special_linear_group n R) : det (g : GL_pos n R).1.1 = 1 :=
+g.prop
```
I think this lemma would be more useful if stated about `general_linear_group.det`, otherwise `simp` can presumably already prove this.

CBirkbeck replied:
Yeah I made it this now.

CBirkbeck committed:
put the lemmas needed here

eric-wieser commented on src/linear_algebra/general_linear_group.lean:
```diff
@@ -220,4 +236,19 @@ end coe_fn_instance
+ext,
+refl,
+end
+
```
These lines are ignoring the comment above that says "-- this section should be last to ensure we do not use it in lemmas"

CBirkbeck replied:
Hmm yes I know, I broke the rules there. The problem is it wont let me state 
```
lemma coe_to_GL_pos_ext (g : special_linear_group n R) (i j : n) :
   (g : (GL_pos n R)) i j = g i j :=
```
before this. Do I need to prove an ext result for `GL_pos` first? Since at the moment it seems to want to make into a function first.

eric-wieser replied:
That's not really an `ext` lemma in the first place - you could spell it
```
example (g : special_linear_group n R) :
  (↑(↑(↑(g : special_linear_group n R) : GL_pos n R) : GL n R) : matrix n n R) i j = (↑g : matrix n n R) := rfl
```
but my claim is that this lemma is useless, it's just a weaker version of the `coe_GL_pos_coe_GL_coe_matrix` we have above.

CBirkbeck committed:
remove bad lemmas

eric-wieser commented on src/linear_algebra/general_linear_group.lean:
```diff
@@ -186,11 +188,25 @@ lemma to_GL_pos_injective :
+  (↑(↑(↑(g : special_linear_group n R) : GL_pos n R) : GL n R) : matrix n n R) = ↑g := rfl
+
+lemma coe_to_GL_pos_det (g : special_linear_group n R) :
+  det ((g : GL_pos n R) : matrix n n R) = (1 : R) :=g.prop
```
Can this be `@[simp]`?

CBirkbeck replied:
Lint complains that simp can already prove this. I'm not sure if this means we shouldn't have this lemma or if it shouldn't be a simp lemma.

eric-wieser replied:
That probably means we shouldn't have this lemma. I pushed a commit that replaces it with one that the linter doesn't complain about

CBirkbeck replied:
I see you added a stronger version of this, thanks!

eric-wieser commented on src/linear_algebra/general_linear_group.lean:
```diff
@@ -186,11 +188,25 @@ lemma to_GL_pos_injective :
+
+variable [fact (even (fintype.card n))]
+
+@[simp] lemma coe_GL_pos_neg (g : special_linear_group n R) : ↑(-g) = -(↑g : GL_pos n R) :=
```
```suggestion
@[simp, norm_cast] lemma coe_GL_pos_neg (g : special_linear_group n R) : ↑(-g) = -(↑g : GL_pos n R) :=
```

CBirkbeck replied:
done

CBirkbeck committed:
add norm cast

eric-wieser commented on src/linear_algebra/general_linear_group.lean:
```diff
@@ -199,8 +199,8 @@ lemma coe_to_GL_pos_det (g : special_linear_group n R) :
-@[simp] lemma coe_GL_pos_neg (g : special_linear_group n R) : ↑(-g) = -(↑g : GL_pos n R) :=
-by {ext, refl}
+@[simp, norm_cast] lemma coe_GL_pos_neg (g : special_linear_group n R) : ↑(-g) = -(↑g : GL_pos n R)
+:= by {ext, refl}
```
Mathlib style never puts `:=` at the start of a line, so you should probably line wrap this after the `:`

CBirkbeck replied:
Oops yes sorry!

eric-wieser committed:
stronger simp lemmas about det

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into SL_to_GL_pos_lemmas

eric-wieser committed:
revert non-terminalization of simp

eric-wieser committed:
style tweak

CBirkbeck committed:
Merge remote-tracking branch 'origin/master' into SL_to_GL_pos_lemmas

CBirkbeck committed:
Merge remote-tracking branch 'origin/master' into SL_to_GL_pos_lemmas

CBirkbeck committed:
lint fix test

riccardobrasca commented on src/linear_algebra/general_linear_group.lean:
```diff
@@ -186,11 +189,26 @@ lemma to_GL_pos_injective :
  from subtype.coe_injective).of_comp
 
+/-- Coercing a `special_linear_group` via `GL_pos` and `GL` is the same as coercing striaght to a
+matrix -/
```
```suggestion
/-- Coercing a `special_linear_group` via `GL_pos` and `GL` is the same as coercing straight to a
matrix. -/
```

riccardobrasca commented on src/linear_algebra/general_linear_group.lean:
```diff
@@ -186,11 +189,26 @@ lemma to_GL_pos_injective :
+matrix -/
+@[simp]
+lemma coe_GL_pos_coe_GL_coe_matrix (g : special_linear_group n R) :
+  (↑(↑(↑(g : special_linear_group n R) : GL_pos n R) : GL n R) : matrix n n R) = ↑g := rfl
```
```suggestion
  (↑(↑(↑g : GL_pos n R) : GL n R) : matrix n n R) = ↑g := rfl
```
Is this necessary?

riccardobrasca commented on src/linear_algebra/general_linear_group.lean:
```diff
@@ -186,11 +189,26 @@ lemma to_GL_pos_injective :
+variable [fact (even (fintype.card n))]
+
+@[norm_cast] lemma coe_GL_pos_neg (g : special_linear_group n R) :
+  ↑(-g) = -(↑g : GL_pos n R) := subtype.ext $ units.ext rfl
```
```suggestion
  ↑(-g) = -(g : GL_pos n R) := subtype.ext $ units.ext rfl
```
I am not totally sure, but if you specify the type there is no need of writing `↑` (similarly elsewhere).

eric-wieser replied:
Sometimes I've had reviewers say they prefer the explicit `↑` as it makes it obvious that a coercion is happening.

riccardobrasca replied:
I am not against it, but we may want to fix a convention... in any case it is not so important.

CBirkbeck replied:
I sort of like having the `↑` but I'm happy to change it if needed.

CBirkbeck committed:
rev fixes

riccardobrasca reviewed (approved):
LGTM.

CBirkbeck committed:
Merge remote-tracking branch 'origin/master' into SL_to_GL_pos_lemmas

jcommelin reviewed (approved):
Thanks :tada:

bors merge
